drivers: flash: Introduce gd32 flash controller driver

cameled opened the pull request:
After read all gd32 series user manual, I get some conclusion for gd32 FMC. But remember, this is not an official document.

This implement GD32 FMC v1, GD32 FMC v2 and GD32F4xx FMC.

## GD32 FMC v1

There have some gd32 FMC only have 1 bank, and their size is smaller than 512K. This group called GD32 FMC v1. 

| GD32 SoC      | Max Flash Size | Bank0 Page Size |
| -----------   | -----------    | -----------     |
| GD32C10x      | 128K           | 1K              |
| GD32C11x      | 128K           | 1K              |
| GD32E10x      | 128K           | 1K              |
| GD32E11x      | 128K           | 1K              |
| GD32E50x      | 512K           | 8K              |
| GD32F3x0      | 128K           | 1K              |
| GD32F1x0      | 64K            | 1K              |
| GD32E23x      | 64K            | 1K              |
| GD32VF103     | 128K           | 1K              |
| GD32L23x      | 256K           | 4/2/1K          |

## GD32 FMC v2

There have some gd32 FMC have 2 bank, and their size can up to 3072K. This group called GD32 FMC v2.

| GD32 SoC    | Max Flash Size | Bank0 Page Size | Bank1 Page Size |
| ----------- | -----------    | -----------     | -----------     |
| GD32A503  | 384K             | 1K              |  1K             |
| GD32F10x    | 3072K          | 2K              | 4K              |
| GD32F20x    | 3072K          | 2K              | 4K              |
| GD32F30x    | 3072K          | 2K              | 4K              |
| GD32F403    | 3072K          | 2K              | 4K              |

For GD32 FMC v2, Bank0 hold the first 512K flash size, and Bank1 for the reset capacity. It have two register set to control Bank0 and Bank1 separately.

In this group, there have some concept called MD(Medium density), HD(High density), XD(Extra density), CL(Connectivity line). Only GD32F10x_MD have some effect for the page size, others can be ignored.

For GD32F10x series, their have some special case for GD32F101x and GD32F103x.

| GD32 SoC    | Flash Size Range | Page Size |
| ---         | ---              | ---       |
| GD32F10x_MD | 16K-256K         | 1K        |
| GD32F10x_HD | 256K-512K        | 2K        |
| GD32F10x_XD | 512K-3072K       | 4K        |

## GD32 FMC v3 (GD32F4xx)

GD32F4xx FMC use `sector` as the minimal operate unit. For GD32F425x, GD32F427x, GD32F470x, `page` are also supported, page size is 4K.

## GD32W51X
GD32W51x FMC is complex, it have FMC mode and QSPI mode. Future developer can pick some to implement, e.g. FMC mode.


## Reference
- [GD32 User Manual](https://gd32mcu.com/en/download/6?kw=)
- [GD32MCU Selection Guide](https://gd32mcu.com/en/download/9?kw=)

de-nordic commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,180 @@
+
+static bool flash_gd32_valid_range(off_t offset, size_t len)
+{
+	if (offset > SOC_NV_FLASH_SIZE) {
```
Shouldn't that be `offset >= SOC_NV_FLASH_SIZE`?
And `off_t` is signed so you may want to check that `offset >= 0`

cameled replied:
Thanks for your suggestion, looks offset like index of an array.
done.

de-nordic commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,180 @@
+		return false;
+	}
+
+	if (len && ((offset + (len - 1)) > SOC_NV_FLASH_SIZE)) {
```
`(offset + len) > SOC_NV_FLASH_SIZE`, for example having `SOC_NV_FLASH_SIZE == 4`, offset == 2 and len == 2 is completely OK.
Should `len == 0` return false?

cameled replied:
> `(offset + len) > SOC_NV_FLASH_SIZE`, for example having `SOC_NV_FLASH_SIZE == 4`, offset == 2 and len == 2 is completely OK.

done.

>Should `len == 0` return false?

I think `len == 0` should return true, there doesn't have any explicit error in here, just no nothing.

de-nordic commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,180 @@
+
+	flash_gd32_unlock();
+
+	for ( ; begin <= end; begin++) {
```
This is my personal opinion: when you are not able to fill in all `for` expressions just use `while`, looks better.

cameled replied:
I think `for` means there have an iteration.

de-nordic commented on drivers/flash/flash_gd32f4xx.c:
```diff
@@ -0,0 +1,167 @@
+	/* Set the PG bit to enable flash programming */
+	FMC_CTL |= FMC_CTL_PG;
+
+	for (size_t i = 0; i < (len / sizeof(flash_prog_t)); i++) {
```
The `flash_prog_t` may have different size, isn't there also alignment requirement to that size?
Shouldn't that be checked by `flash_gd32_valid_range maybe`?

cameled replied:
done.

de-nordic commented on drivers/flash/flash_gd32f4xx.c:
```diff
@@ -0,0 +1,167 @@
+		/* Nop */
+	}
+
+	if (FMC_STAT & FMC_STAT_OPERR) {
+		flash_gd32_err_handler();
+		return -EIO;
+	}
+
+	/* Clear the PG bit to disable flash programming */
+	FMC_CTL &= ~FMC_CTL_PG;
```
Even though error happened, shouldn't the `FMC_CTL_PG` bit be cleared anyway?

cameled replied:
Thanks! I missed the safe return check.
done.

de-nordic commented on drivers/flash/flash_gd32f4xx.c:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	/* Wait for the sector erase complete. */
+	while (FMC_STAT & FMC_STAT_BUSY) {
+		/* Nop */
+	}
```
The `flash_gd32_page_erase` checks for timeout on similar loop, shouldn't this be also happening here?

cameled replied:
Add a timeout in here.

cameled commented:
@nandojve @gmarull Just implement a fmc driver for gd32f4xx series.

cameled commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,199 @@
+	.erase_value = 0xff,
+};
+
+__weak int flash_gd32_programming(off_t offset, const void *data, size_t len)
```
@nandojve Please add a DMN for this PR.
I use a weak implement to avoid compile time error. But I think there should have a better solution.

nandojve replied:
You can tell people to not merge by adding a DNM at PR.

nandojve replied:
You can drop all weak functions.

cameled replied:
use GD32_HAS_FLASH_DRIVER, done.

nandojve commented on drivers/flash/Kconfig.gd32:
```diff
@@ -0,0 +1,12 @@
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_HAS_DRIVER_ENABLED
+	select MPU_ALLOW_FLASH_WRITE if ARM_MPU
+	depends on (SOC_FAMILY_GD32 || SOC_SERIES_GD32VF103)
```
My suggestion is:

Add `GD32_HAS_FLASH_DRIVER` at modules/hal_gigadevice

Check by `GD32_HAS_FLASH_DRIVER` instead SoC Family or Series.
```diff
- depends on (SOC_FAMILY_GD32 || SOC_SERIES_GD32VF103)
+ depends on GD32_HAS_FLASH_DRIVER
```

Than add `GD32_HAS_FLASH_DRIVER` at SoC https://github.com/zephyrproject-rtos/zephyr/blob/bc72fb67b1c81670a087f766ef38c3517da10366/soc/arm/gigadevice/gd32f4xx/Kconfig.series#L11-L12

This way flash driver will be only available for those SoC that have a driver.

cameled replied:
done.

nandojve commented on drivers/flash/CMakeLists.txt:
```diff
@@ -56,6 +56,13 @@ if(CONFIG_SOC_FLASH_STM32)
+  zephyr_library_sources(flash_gd32.c)
+
+  zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_GD32F4XX flash_gd32f4xx.c)
+  zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_GD32F3X0 flash_gd32f3x0.c)
```
Do you have this driver right now? Otherwise you should drop it:

```diff
-zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_GD32F3X0 flash_gd32f3x0.c)
```

cameled replied:
done.

cameled commented:
I will continue this work recently, final result will be a series general FMC drivers for gd32.

cameled commented:
| GD32 SoC      | Page Size |
| ----------- | ----------- |
| GD32C10x      | 1K       |
| GD32C11x      | 1K       |
| GD32E10x      | 1K       |
| GD32E10x      | 1K       |

nandojve commented:
Hi @cameled ,

Thank you for your effort. Can you explain in few words why API is so different?

cameled commented:
> Hi @cameled ,
> 
> Thank you for your effort. Can you explain in few words why API is so different?

I updated the original description. May help developer to understand GD32 FMC.

nandojve commented:
Hi @cameled ,
I would like know if you have time to address my comments from March. It will be nice if we can have this driver for v3.2.

cameled commented:
> Hi @cameled , I would like know if you have time to address my comments from March. It will be nice if we can have this driver for v3.2.

Sorry for the long delay, I recently struggled with yocto, our new project use a custom SDK based on it.

Anyway, I will compelete this on  Thuesday, For v1, v2 and GD32F4XX. And use `GD32_HAS_FLASH_DRIVER` to allow work with GD32W51X.

Commit:
dts: add gd32 fmc flash memory info

There are three types GD32 FMC.

GD32 FMC v1: its flash memory has 1 bank, page size is equal in the
bank, flash size is smaller than 512KB.

GD32 FMC v2: its flash memory has 2 banks. Page size equal within the
same bank but different between banks. Flash size can be up to 3072KB.
FMC v2 has two registers to control bank0 and bank1 separately.

GD32 FMC v3: its flash memory has 2 banks, use sector size as the
minimum operating unit, the sector size is not equal.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
drivers: flash: introduce gd32 fmc driver

This supports three types GD32 FMC flash memory. GD32 FMC v1,
GD32 FMC v2 and GD32 FMC v3.

GD32 FMC v1 for small flash memory, flash size can be up to 512KB.

GD32 FMC v2 for large flash memory, flash size can be up to 3072KB.

GD32 FMC v3 not use page but sector as minimum block, flash size can
be up to 3072KB.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
boards: gd32 board add zephyr,flash-controller property

This add zephyr,flash-controller property for supported boards.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
samples: flash_shell: add support for gd32 boards

After add zephyr,flash-controller property, most gd32 boards support
flash_shell sample.

gd32vf103c_starter and gd32vf103v_eval only have 32KB SRAM, so we
should reduce CONFIG_HEAP_MEM_POOL_SIZE to 8KB.

gd32f350r_eval only have 16KB driver, exclude from flash_shell sample.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

cameled commented:
Test on `gd32f450i_eval`(GD32F4xx FMC) and `gd32vf103v_eval`(GD32 FMC v1) boards. 

GD32 FMC v2 for `gd32f403z_eval` board build success, but't test on board.

Others GD32 FMC v1 boards also build success. (gd32f350r_eval not support, memory to small)

cameled commented on drivers/flash/flash_gd32_v1.c:
```diff
@@ -0,0 +1,198 @@
+};
+#endif
+
+static inline void flash_gd32_fmc_v1_unlock(void)
```
`sector` and `page` are different concept. I think it's more clearly to split them.

cameled commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,121 @@
+		return -EINVAL;
+	}
+
+	if (!len) {
```
duplicated.

cameled commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,121 @@
+		return -EINVAL;
+	}
+
+	if (!len) {
```
Done, more easily to catch the exit reason, nothing need to do.

cameled commented on drivers/flash/flash_gd32_v1.c:
```diff
@@ -0,0 +1,198 @@
+	uint32_t page_addr = FLASH_GD32_ADDR + offset;
+	int ret = 0;
+
+	while (size) {
```
I think there use `while (size) {` is ok, do it while `size` is available.

nandojve commented on drivers/flash/Kconfig.gd32:
```diff
@@ -0,0 +1,29 @@
+	depends on GD32_HAS_FLASH_DRIVER
+	select FLASH_HAS_DRIVER_ENABLED
+	default y
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32E10X
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32E50X
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32F3X0
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32VF103
+	select SOC_FLASH_GD32_V2 if SOC_SERIES_GD32F403
```
The **SOC_FLASH_GD32_V1/V2** should be selected directly at SoC KConfig

nandojve reviewed (changes requested):
Overall LGTM, just minors.

nandojve commented:
@gmarull @de-nordic , do you mind revisit this PR? I think is ready to go.

gmarull commented on drivers/flash/flash_gd32.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef ZEPHYR_DRIVERS_FLASH_FLASH_GD32_H_
+#define ZEPHYR_DRIVERS_FLASH_FLASH_GD32_H_
+
+#include <errno.h>
```
doesn't seem we use errno here, but we need `<stdint.h>`

cameled replied:
done.

gmarull commented on drivers/flash/flash_gd32_v1.c:
```diff
@@ -0,0 +1,198 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gd32_fmc.h>
+
+#include "flash_gd32.h"
+
+#include <zephyr/logging/log.h>
```
nit:

```
#include "flash_gd32.h"

#include <zephyr/logging/log.h>

#include <gd32_fmc.h>
```

cameled replied:
done.

gmarull commented on drivers/flash/flash_gd32_v1.c:
```diff
@@ -0,0 +1,198 @@
+
+	while (FMC_STAT & FMC_STAT_BUSY) {
+		if (k_uptime_get() > expired_time) {
+			return -EIO;
```
`-ETIMEDOUT`?

cameled replied:
user `-ETIMEDOUT`, done.

gmarull commented on drivers/flash/flash_gd32_v2.c:
```diff
@@ -0,0 +1,415 @@
+	uint32_t page_addr = FLASH_GD32_ADDR + offset;
+	int ret = 0;
+
+	while (size) {
```
nit: use boolean expression here `> 0U`

gmarull commented on drivers/flash/flash_gd32f4xx.c:
```diff
@@ -0,0 +1,263 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gd32_fmc.h>
+
+#include "flash_gd32.h"
+
+#include <zephyr/logging/log.h>
```
ditto

gmarull commented on drivers/flash/flash_gd32f4xx.c:
```diff
@@ -0,0 +1,263 @@
+#endif
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+/* SN bits in FMC_CTL are not continues value, use below table to mapping it. */
```
`/* SN bits in FMC_CTL are not continue values, use table below to map them. */`

cameled replied:
Thanks for your advice! done.

gmarull commented on drivers/flash/flash_gd32f4xx.c:
```diff
@@ -0,0 +1,263 @@
+
+	while (FMC_STAT & FMC_STAT_BUSY) {
+		if (k_uptime_get() > expired_time) {
+			return -EIO;
```
`-ETIMEDOUT`?

gmarull commented on samples/drivers/flash_shell/boards/gd32vf103c_starter.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_HEAP_MEM_POOL_SIZE=8192
```
nit: add a comment explaining why (same as explained in the commit message)

cameled replied:
Add a comment to this config, user can easily to know why this is needed.

gmarull replied:
can't see the comment, missed push?

cameled replied:
Sorry, forget to apply this patch. done.

gmarull reviewed (commented):
overall lgtm, a few comments/suggestions

fkokosinski commented on drivers/flash/flash_gd32.h:
```diff
@@ -0,0 +1,44 @@
+#endif
+
+/* Helper for conditional compilation directives, KB have a type casting. */
+#define PRE_KB(x) ((x) << 10)
```
Use `KB` macro from `sys/util.h`.

cameled replied:
`KB` can not be used for preprocessor math operation, it have a type casting `(size_t)`.

fkokosinski commented on drivers/flash/flash_gd32_v1.c:
```diff
@@ -0,0 +1,198 @@
+	uint32_t page_addr = FLASH_GD32_ADDR + offset;
+	int ret = 0;
+
+	while (size) {
```
What if `size` gets below 0 because of setting/passing wrong values here?

cameled replied:
Thanks for point out this, @soburi first mention this for me, but I'm not sure their have any requirement for this. Anyway, change to use this type.

fkokosinski commented on dts/bindings/mtd/gd,gd32-nv-flash.yaml:
```diff
@@ -0,0 +1,31 @@
+
+  bank0-page-size:
+    type: int
+    required: false
```
IIRC properties aren't required by default, no point in setting it here.

fkokosinski commented on dts/bindings/mtd/gd,gd32-nv-flash.yaml:
```diff
@@ -0,0 +1,31 @@
+
+  bank1-page-size:
+    type: int
+    required: false
```
Ditto

fkokosinski commented on modules/hal_gigadevice/Kconfig:
```diff
@@ -14,6 +14,11 @@ config GD32_HAS_AFIO_PINMUX
+config GD32_HAS_FLASH_DRIVER
+	bool
+	help
+	  This option should be selected if GD32 FMC driver support the series.
```
`supports`

cameled commented on drivers/flash/Kconfig.gd32:
```diff
@@ -0,0 +1,29 @@
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32E50X
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32F3X0
+	select SOC_FLASH_GD32_V1 if SOC_SERIES_GD32VF103
+	select SOC_FLASH_GD32_V2 if SOC_SERIES_GD32F403
```
Add `gd,gd32-nv-flash-v1`, `gd,gd32-nv-flash-v2` and `gd,gd32-nv-flash-v3`(gd32f4xx).

gd32 flash driver will only rely on `gd,gd32-nv-flash-vx` to enable it by default.

cameled commented on drivers/flash/CMakeLists.txt:
```diff
@@ -61,6 +61,14 @@ if(CONFIG_SOC_FLASH_STM32)
   endif()
 endif()
 
+if(CONFIG_SOC_FLASH_GD32)
```
A little different in here, gd32 only have one driver api implemnt, located in `flash_gd32.c`. For real flash memory operation, there have three sub-driver `v1`, `v2` and `v3`.

cameled commented:
I adjust the flash_gd32 driver enable hehavior. After add `gd,gd32-nv-flash-vx` to flash node compatible, flash_driver will enable be default.

For example, after add `gd,gd32-nv-flash-v3`, flash_gd32 driver will enable by default. It will use `gd32-nv-flash-v3` flash memory sub-driver.
 
```
/* gd32f4xx.dtsi */
			flash0: flash@8000000 {
				compatible = "gd,gd32-nv-flash-v3", "soc-nv-flash";
				write-block-size = <2>;
				max-erase-time = <8000>;
			};
```

If new added GD32 FMC not supported by flash_gd32 driver, Use  "soc-nv-flash" to provide basic flash memory infomation to Zephyr.
```
/* gd32f4xx.dtsi */
			flash0: flash@8000000 {
				compatible = "soc-nv-flash";
				write-block-size = <2>;
				max-erase-time = <8000>;
			};
```

gmarull commented on drivers/flash/flash_gd32_v2.c:
```diff
@@ -0,0 +1,418 @@
+#define GD32_NV_FLASH_V2_NODE		DT_INST(0, gd_gd32_nv_flash_v2)
+#define GD32_NV_FLASH_V2_TIMEOUT	DT_PROP(GD32_NV_FLASH_V2_NODE, max_erase_time)
+
+/**
+ * @brief GD32 FMC v2 flash memory have 2 bank.
+ * Bank0 hold the first 512KB, bank1 for the reset capacity.
+ * The page size is same in one bank, but different between banks.
+ */
```
```suggestion
/**
 * @brief GD32 FMC v2 flash memory has 2 banks.
 * Bank0 holds the first 512KB, bank1 is used give capacity for reset.
 * The page size is the same within the same bank, but not equal for all banks.
 */
```

cameled replied:
Thanks! applied.

gmarull commented on drivers/flash/flash_gd32_v3.c:
```diff
@@ -0,0 +1,261 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(flash_gd32);
+
+#include "flash_gd32.h"
+
+#include <gd32_fmc.h>
```
```suggestion
#include "flash_gd32.h"

#include <zephyr/logging/log.h>

#include <gd32_fmc.h>

LOG_MODULE_DECLARE(flash_gd32);
```

cameled replied:
applied.

gmarull commented on dts/bindings/flash_controller/gd,gd32-flash-controller.yaml:
```diff
@@ -1,7 +1,18 @@
 
-description: GD32 flash controller
+description: |
+  There have three types GD32 FMC.
+
+  GD32 FMC v1, it flash memory have 1 bank, page size is same in bank,
+  flash size smaller than 512KB.
+
+  GD32 FMC v2, it flash memory has 2 banks. Page size is same in bank but
+  different between banks. Flash size can up to 3072KB. FMC v2 have two
+  register set to control bank0 and bank1 separately.
+
+  GD32 FMC v3, it flash memory has 2 bank, use sector as minimum operating
+  unit, sector size is not same.
```
```suggestion
  There are three types GD32 FMC.

  GD32 FMC v1: its flash memory has 1 bank, page size is equal in the bank,
  flash size is smaller than 512KB.

  GD32 FMC v2: its flash memory has 2 banks. Page size equal within the same bank but
  different between banks. Flash size can be up to 3072KB. FMC v2 has two
  registers to control bank0 and bank1 separately.

  GD32 FMC v3: its flash memory has 2 banks, use sector size as the minimum operating
  unit, the sector size is not equal.
```

cameled replied:
Thanks for your suggestion, applied.

gmarull commented on dts/bindings/mtd/gd,gd32-nv-flash-v1.yaml:
```diff
@@ -0,0 +1,32 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Flash memory binding of gd32 fmc v1.
```
```suggestion
  Flash memory binding of GD32 FMC v1.
```

gmarull commented on dts/bindings/mtd/gd,gd32-nv-flash-v1.yaml:
```diff
@@ -0,0 +1,32 @@
+compatible: gd,gd32-nv-flash-v1
+
+properties:
+  max-erase-time:
```
```suggestion
  max-erase-time-ms:
```

(add units)

cameled replied:
applied, add uint `ms` to `v1`, `v2` and `v3` max erase time property.

gmarull reviewed (commented):
lgtm, a few last nitpicks

cameled commented on drivers/flash/flash_gd32.c:
```diff
@@ -0,0 +1,121 @@
+#include "flash_gd32.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(flash_gd32, CONFIG_FLASH_LOG_LEVEL);
```
Thanks for you suggestion, applied.

cameled commented on drivers/flash/flash_gd32.h:
```diff
@@ -0,0 +1,43 @@
+#error "Invalid write-block-size value in FMC DTS"
+#endif
+
+/* Helper for conditional compilation directives, KB have a type casting. */
```
applied.

cameled commented on drivers/flash/flash_gd32_v1.c:
```diff
@@ -0,0 +1,196 @@
+
+#include "flash_gd32.h"
+
+#include <gd32_fmc.h>
```
applied.
